consul/agents.go

```diff
@@ -47,18 +51,45 @@ func (a *ConcurrentAgents) getRandomAgentHost() string {
 	return hosts[idx]
 }
 
-func (a *ConcurrentAgents) GetAgent(agentHost string) (*consulapi.Client, error) {
+func (a *ConcurrentAgents) RemoveAgent(agentAddress string) {
 	a.lock.Lock()
 	defer a.lock.Unlock()
-	if agent, ok := a.agents[agentHost]; ok {
+
+	if IP, err := utils.HostToIPv4(agentAddress); err != nil {
+		log.WithError(err).Error("Could not remove agent from cache")
+	} else {
+		agentAddress = IP.String()
+		if _, ok := a.agents[agentAddress]; ok {
+			log.WithField("Address", agentAddress).Info("Removing agent from cache")
+			delete(a.agents, agentAddress)
+			a.updateAgentsCacheSizeMetricValue()
+		} else {
+			log.WithField("Address", agentAddress).Info("Agent not found in cache")
+		}
+	}
+}
+
+func (a *ConcurrentAgents) GetAgent(agentAddress string) (*consulapi.Client, error) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	IP, err := utils.HostToIPv4(agentAddress)
+	if err != nil {
+		return nil, err
+	}
+	agentAddress = IP.String()
```

Comment:
Maybe `HostToIPv4` should return string or (better idea) we should use `IP` everywhere we can to have clear situation what is stored in `agentAddress`. Another thing overriding parameters value is considered as a bad idea, it is difficult to read because `agentAddress` we are using here is not `agentAddress` that was passed